### PR TITLE
nixos/snapraid: init

### PR DESCRIPTION
**Pull request by lunik1**
Hello maintainers, you can just skip to [this comment](https://github.com/NixOS/nixpkgs/pull/127245#issuecomment-864556034) for a more refined description and summary.

###### Motivation for this change

Adds a snapraid module to NixOS

A draft for now since there a few things I'm not happy with:
- [ ] It might be in the wrong directory
- [x] I haven't tested it
- [x] Snapraid can only perform one operation at a time but this isn't enforced by this module. This might require some clever (impossible?) systemd tricks to make e.g. sync wait for scrub to finish
- [ ] The config generation is a little arcane
- [ ] The choice of IO priority is a bit arbitrary - maybe it should be configurable?
- [ ] No drive pooling support outside of extraConfig, but most users (including myself) will use mergerfs amyway

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by lunik1** [comment]
`snapraid-scrub` will now wait for `snapraid-sync` to finish

**Comment by lunik1** [comment]
Ok I think this is a mergeable state now, although I still have a few concerns maybe the reviewers can help with:

This PR adds a [SnapRAID](https://snapraid.it) module to NixOS. It will run `snapraid scrub` and `snapraid sync` at user-defined intervals, and if there is a sync in progress a scrub will wait for it to finish (but not vice versa). The amount and age of files scrubbed is configurable. By default `snapraid touch` is used to set the sub-second timestamp before each sync.

The two tasks are run at low i/o and CPU priority, and have various hardening options mostly inspired by [borgmatic](https://github.com/witten/borgmatic/blob/b37dd1a79e5a992325080f232ebd039be5742792/sample/systemd/borgmatic.service).

This module is also able to generate the `/etc/snapraid.conf` file.

Possible issues:
- I'm not sure `modules/tasks` is the best folder, `services/backup` might be more appropriate?
- `snapraid-sync` will fail if `snapraid-scrub` is running, but it should be more common to run sync before scrub
- The i/o and CPU are set to the lowest non-idle priority, but perhaps they should be little higher so there is still room for users to run processes at an even lower priority (or it should be configurable?)
- I think the config generation code may be a little unclear, maybe some more experienced Nix users can chime in with possible improvements

**Comment by makefu** [comment]
Hey @lunik1 first of all, thanks for your effort brining this module upstream! I will try to look into your module this evening but it sure looks promising!

Regarding Parity and Content Files, Snapraid gives some insight in the FAQ (https://www.snapraid.it/faq#fcontent), it might be a good idea to reflect this info in the module description (e.g. that the parity file will be as big as the largest disk in your array and that the content files can reside on multiple places (not only the parity disk) to provide redundancy of the information.
We could even think of setting the contentFiles by default to reside on all raid disks plus /var/ or something, just like you write in the example.

Anyway this is not mandatory for this PR to be merged, just would be a cool addition

**Comment by lunik1** [comment]
I tried to describe options in a way that would make sense if you were already familiar with SnapRAID and were just looking for the equivalent Nix option. I can see how it would make sense to at least point in the direction of the relevant parts of the SnapRAID docs, though.

**Comment by makefu** [comment]
Hey @lunik1! I found some time to try your PR.

i tried snapraid with [nixos-shell]( https://github.com/Mic92/nixos-shell ) with the following config:
```nix
{ pkgs, ... }: {
  boot.kernelPackages = pkgs.linuxPackages_latest;
  system.activationScripts.init_disks = ''
    mkdir -p /media/{1,2,3,par}

    echo "ONE" > /media/1/first_file
    echo "TWO" > /media/2/second_file
    echo "THREE" > /media/3/third_file
  '';
  services.openssh.enable = true;
  snapraid ={
    enable = true;
    dataDisks = {
      one = "/media/1";
      two = "/media/2";
      three = "/media/3";
    };
    parityFiles = [
      "/media/par/par.parity"
    ];
    contentFiles = [
      "/media/1/1.content"
      "/media/2/2.content"
      "/media/3/3.content"
      "/media/par/par.content"
    ];
  };
}
```
however when trying your patch onto the latest master and setting NIX_PATH to this checkout i get the following error:
```bash
root@nixos> systemctl start snapraid-sync.service
Job for snapraid-sync.service failed because the control process exited with error code.
See "systemctl status snapraid-sync.service" and "journalctl -xe" for details.
root@nixos> systemctl status snapraid-sync
● snapraid-sync.service - Synchronize the state of the SnapRAID array
     Loaded: loaded (/nix/store/wq12yh9fxppa41lszc23aqyk2qp6589z-unit-snapraid-sync.service/snapraid-sync.service; linked; vendor preset: enabled)
     Active: failed (Result: exit-code) since Sun 2021-06-20 21:37:40 UTC; 31s ago
TriggeredBy: ● snapraid-sync.timer
    Process: 941 ExecStartPre=/nix/store/nn4g4glggacm3bw4bjndgz8kbhyxkfp1-snapraid-11.5/bin/snapraid touch (code=exited, status=226/NAMESPACE)
         IP: 0B in, 0B out
        CPU: 3ms

Jun 20 21:37:40 nixos systemd[1]: Starting Synchronize the state of the SnapRAID array...
Jun 20 21:37:40 nixos systemd[941]: snapraid-sync.service: Failed to set up mount namespacing: /run/systemd/unit-root/media/par/par.content: No such file or directory
Jun 20 21:37:40 nixos systemd[941]: snapraid-sync.service: Failed at step NAMESPACE spawning /nix/store/nn4g4glggacm3bw4bjndgz8kbhyxkfp1-snapraid-11.5/bin/snapraid: No such file or directory
Jun 20 21:37:40 nixos systemd[1]: snapraid-sync.service: Control process exited, code=exited, status=226/NAMESPACE
Jun 20 21:37:40 nixos systemd[1]: snapraid-sync.service: Failed with result 'exit-code'.
Jun 20 21:37:40 nixos systemd[1]: Failed to start Synchronize the state of the SnapRAID array.
```

am i missing something?

**Comment by lunik1** [comment]
I can repro, It seems that with the current permissions SnapRAID cannot create the content and parity files, but will proceed as normal if they already exist. Verified by adding 
```
    touch /media/par/par.content
    touch /media/par/par.parity
    touch /media/1/1.content
    touch /media/2/2.content
    touch /media/3/3.content
```
to the init script in your config, which lets us get as far as
```
Jun 20 22:22:36 nixos snapraid[866]: Disks '/media/1/' and '/media/3/' are on the same device.
```

I assume this is because of the service hardening, I'll try to find the culprit.

**Comment by lunik1** [comment]
Latest push _should_ correct this. Fairly simple fix in the end: just preface the ReadWritePaths with `-` so `systemd` doesn't error if they don't exist.

**Comment by makefu** [comment]
hey, i've tested with the following configuration:

```nix
{ pkgs, ... }: {
  boot.kernelPackages = pkgs.linuxPackages_latest;
  # for some reason i get operation not permitted ...
  system.activationScripts.init_disks = ''
    echo " writing do_test.sh"
    cat <<EOF >/root/do_test.sh
    #!/bin/sh
    set -x
    mkdir -p /media/{1,2,3,par}
    for i in 1 2 3 par;do
      truncate -s 1M /\$i.ext4
      ${pkgs.e2fsprogs}/bin/mkfs.ext4 -F /\$i.ext4
      echo mount -text4 -o loop /\$i.ext4 /media/\$i
      mount -text4 -o loop /\$i.ext4 /media/\$i
    done

    echo "ONE" > /media/1/first_file
    echo "TWO" > /media/2/second_file
    echo "THREE" > /media/3/third_file
    EOF

  '';
  services.openssh.enable = true;
  snapraid ={
    enable = true;
    dataDisks = {
      one = "/media/1";
      two = "/media/2";
      three = "/media/3";
    };
    parityFiles = [
      "/media/par/par.parity"
    ];
    contentFiles = [
      "/media/1/1.content"
      "/media/2/2.content"
      "/media/3/3.content"
      "/media/par/par.content"
    ];
  };
}
```
however i am getting now the following error after running the following commands:
```
root@nixos> sh /root/do_test.sh                                                                                                                        ~/r/nixos-shell
...
root@nixos> systemctl start snapraid-sync                                                                                                              ~/r/nixos-shell
Job for snapraid-sync.service failed because the control process exited with error code.
See "systemctl status snapraid-sync.service" and "journalctl -xe" for details.
[1] root@nixos> systemctl status snlastapraid-sync | ncat                                                                                                  ~/r/nixos-shell
● snapraid-sync.service - Synchronize the state of the SnapRAID array
     Loaded: loaded (/nix/store/54mi7vbcmvn8mpi4wscd52s7m9ysvx8k-unit-snapraid-sync.service/snapraid-sync.service; linked; vendor preset: enabled)
     Active: failed (Result: exit-code) since Mon 2021-06-21 09:19:50 UTC; 17s ago
TriggeredBy: ● snapraid-sync.timer
    Process: 891 ExecStartPre=/nix/store/nn4g4glggacm3bw4bjndgz8kbhyxkfp1-snapraid-11.5/bin/snapraid touch (code=exited, status=1/FAILURE)
         IP: 0B in, 0B out
        CPU: 28ms

Jun 21 09:19:50 nixos snapraid[891]: No content file found. Assuming empty.
Jun 21 09:19:50 nixos snapraid[891]: Setting sub-second timestamps...
Jun 21 09:19:50 nixos snapraid[891]: Saving state to /media/1/1.content...
Jun 21 09:19:50 nixos snapraid[891]: Saving state to /media/2/2.content...
Jun 21 09:19:50 nixos snapraid[891]: Saving state to /media/3/3.content...
Jun 21 09:19:50 nixos snapraid[891]: Saving state to /media/par/par.content...
Jun 21 09:19:50 nixos snapraid[891]: Error removing the stale content file '/media/par/par.content.tmp'. Read-only file system.
Jun 21 09:19:50 nixos systemd[1]: snapraid-sync.service: Control process exited, code=exited, status=1/FAILURE
Jun 21 09:19:50 nixos systemd[1]: snapraid-sync.service: Failed with result 'exit-code'.
Jun 21 09:19:50 nixos systemd[1]: Failed to start Synchronize the state of the SnapRAID array.
```

**Comment by lunik1** [comment]
Hmm ok what seems to be happening here is that if `ReadWritePaths` specifies a path to a file you can read and write to it, but not delete it. This requires write access to the directory above. Looking into a fix.

**Comment by lunik1** [comment]
Fixed by adding the directories containing the contents files to `ReadWritePaths`.

I did notice that with your VM spec that `snapraid touch` didn't seem to set the sub-second timestamps, but when using XFS instead of ext4 I didn't get the same issue. Not sure what's going on but I don't think it is connected to SnapRAID.

**Review by makefu — APPROVED** [review]
PR looks good to me now! I managed to get everything running successfully with the configuration in the thread.
thanks for the efforts!

**Comment by lunik1** [comment]
I don't know why the manual tests have started failing...

**Comment by ryantm** [comment]
> I don't know why the manual tests have started failing...

You may need to rebase this on master and re-run `nixos/doc/manual/md-to-db.sh`.

**Comment by lunik1** [comment]
Using that script is much easier than editing the xml by hand 😅. Looks like we are good to go now?

**Comment by lunik1** [comment]
@ryantm are there any blockers on this other than resolving the merge conflict on the release notes? I would like to merge this this week if possible, will be slower to respond after that for the foreseeable future.

File: nixos/modules/tasks/snapraid.nix
Comment (ryantm) [inline nixos/modules/tasks/snapraid.nix:147]: You could use `systemd.services.<name>.startAt` instead of explicitly specifying the timers here.

https://search.nixos.org/options?channel=21.05&show=systemd.services.%3Cname%3E.startAt&from=0&size=50&sort=relevance&query=systemd.services.%3Cname%3E

File: nixos/doc/manual/release-notes/rl-2111.section.md
Comment (ryantm) [inline nixos/doc/manual/release-notes/rl-2111.section.md:22]:
```suggestion
  Available as [snapraid](#opt-snapraid.enable).
```
I believe options.html is not necessary here.

File: nixos/modules/tasks/snapraid.nix
Comment (ryantm) [inline nixos/modules/tasks/snapraid.nix:223]: I don't know anything about SnapRAID but are you sure  `/var` and `/media/par` should be here and not in some configuration option that can be overridden?

**Review by ryantm — CHANGES_REQUESTED** [review]
Without knowing much about SnapRAID, this looks okay to me. I had a few comments to review.

File: nixos/modules/tasks/snapraid.nix
Comment (lunik1) [reply nixos/modules/tasks/snapraid.nix]: You're right, `/media/par` was leftover. I thought `/var/` was required as that's where SnapRAID appeared to keep its lockfile, but on closer inspection it actually uses the directory of the first content file.

**Review by ryantm — APPROVED** [review]
I haven't tested this but it looks good to me. If @lunik1 is happy with it, I'm happy to merge.

**Comment by lunik1** [comment]
I'm happy to merge too 🎉.

**Comment by makefu** [comment]
+1 and like from my side as well. Thanks for your effort!

**Comment by makefu** [comment]
:tada:

**Pull request merged**
